Fix use of IntelliJ IDEA dependencies scopes.

nskvortsov opened the pull request:
Reduce number of dependencies in generated or imported Idea projects by using the "scope inheritance" similar to what Gradle have. E.g., compile and provided scope dependencies are always included in test compile and runtime, etc.

This reduces number of dependencies twice on average, speeding up project import, sync and improving overall IntelliJ performance on middle to large projects.

oehme commented:
I'd like to stay away from this `plus/minus` logic as much as possible. We simplified this in Gradle 3.4 and it fixed a couple of bugs in the process. It would be great if we didn't reintroduce those. Running `ide:check` should give you a good idea of what use cases would fail with this change.

How big is the performance impact of this? Can we do something that doesn't complicate our mapping again?

nskvortsov commented:
@oehme thank you for reviewing this.
Import/sync operation duration (outside the Gradle process) is halved with this patch, reducing overall timing by approx. 20% for me. Beside that, suggested mapping creates much more natural dependencies list, much closer to what a user would set up manually if project is configured from scratch.

The dependencies mapping is not very complicated with suggested syntax and is a good trade-off for performance and natural dependencies presentation.

oehme commented:
Can you please run `ide:check`? You may see some breaking changes (like two different versions of the same module ending up on the classpath). If nothing except the scope itself changes I'd be fine with this.

Please don't reference the `compile` configuration anywhere. We should make as few assumptions about the shape of the configuration hierarchy as possible. Use `compileClasspath` and `runtimeClasspath` wherever you can.

nskvortsov commented:
Thank you. I ran `ide:check` and updated several tests in this pull request. Still working on `ToolingApiIdeaModelCrossVersionSpec`. Will review the `compile` configuration usage after that

nskvortsov commented:
I've removed usage of deprecated configuration name (sadly, with penalty in performance gain). Also fixed `ToolingApiIdeaModelCrossVersionSpec` as far as I could. Still, I do not know how to deal with test variant that executes build with previous version (v. 4.3) and presents (wrong) results to test assertions.

oehme commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/IdeaPlugin.java:
```diff
@@ -391,42 +391,34 @@ public FileCollection call() throws Exception {
+        scopes.named(GeneratedIdeaScope.TEST)
+            .plus(JavaPlugin.TEST_COMPILE_CLASSPATH_CONFIGURATION_NAME)
+            .plus(JavaPlugin.TEST_RUNTIME_CLASSPATH_CONFIGURATION_NAME)
+            .minus(JavaPlugin.COMPILE_CLASSPATH_CONFIGURATION_NAME);
```
Does TEST in idea also extend RUNTIME? If so, you could also do `minus(RUNTIME_CLASSPATH)` here.

nskvortsov replied:
Idea's scope do not extend each other, but rather contribute to a classpath. To be exact, RUNTIME scope deps are added to _tests execution_ classpath.  `minus(RUNTIME_CLASSPATH)` will remove duplicate deps coming from `runtimeOnly` but that's not a big number usually.

oehme replied:
It's still something, I'd add it :)

oehme commented:
The way to deal with cross-version tests is to add an upper bound for the old behavior and add a new test for the new behavior with the lower bound being the current version.

nskvortsov commented:
Thank you, `ToolingApiIdeaModelCrossVersionSpec` works now. Moving on to `integTest` now as I totally forgot about Gradle's `--continue` flag :)

oehme commented:
I thought about this some more and I think we can get the full benefit without making the mapping step more complex. Instead, we could do some post-processing on the IDEA scopes, using our knowledge of IDEA:

- if a dependency is in COMPILE, remove it from PROVIDED, RUNTIME, TEST
- if a dependency is in PROVIDED or RUNTIME, remove it from TEST
- if a dependency is in PROVIDED **and** RUNTIME, move it up to COMPILE

That way you get the minimal scopes you were aiming for while avoiding a complex plus/minus mapping from Gradle's Configurations. Does that sounds reasonable?

nskvortsov commented:
I would like to give it a try (at least). While working on this pull request's details I found that unfortunately, there is no proper way to map Gradle's  production/test + compile/runtime classpaths to IJ IDEA's scopes with only "plus/minus" notation.

Post-processing can be a good solution if you are fine with separating the setup of defaults and optimizing them to proper state. I will start looking into it in another branch

nskvortsov commented:
Unfortunately, you can not remove RUNTIME scope deps from TEST scope, as you can accidentally remote something included by "testCompileOnly" scope...

oehme commented:
That's true. Fortunately this will only affect `runtimeOnly` dependencies, so the performance impact will be small. The other rules outlined above should work.

nskvortsov commented:
I have implemented the post-processing with optimization and submitted it as a separate pull request.
Please, take a look at it: https://github.com/gradle/gradle/pull/3460

oehme commented:
We can close this one then, as #3460 has the same effect (and more).
